### ISSUE TITLE
Integrated Terminal fails to display color prompt on Debian, Ubuntu, Mint

### ISSUE THREAD

**Issue by rmunn** [Microsoft/vscode#9681]
- VSCode Version: 1.3.1 (Commit e6b4afa53e9c0f54edef1673de9001e9f0f547ae)
- OS Version: Ubuntu 14.04 (Trusty), Ubuntu 16.04 (Xenial), Debian 8.4 (Jessie), Linux Mint 18 (Sarah)

## Steps to Reproduce

1. Open VS Code.
2. Press `Ctrl-backtick` to open integrated terminal.
3. Observe whether the terminal prompt is colored or not.

## Details

On Debian-based systems (Debian, Ubuntu, Linux Mint), the default `.bashrc` in users' home folders sets the prompt to `(username)@(hostname):(current working directory)$(space)`. Additionally, if the terminal is capable of displaying color, it uses terminal escapes to color the username and hostname green, and the current working directory blue. The way the default `.bashrc` detects a color-capable terminal is by looking at the value of the `$TERM` environment variable; if it is `xterm-color` or `xterm-256color`, the color prompt is used; otherwise, a black&white prompt is used.

And this is where VS Code comes into the picture. In [line 18 of `terminalProcess.js`](https://github.com/Microsoft/vscode/blob/9a7ec35a6adc97b7b611a9bdc6228eaea7c6be0d/src/vs/workbench/parts/terminal/electron-browser/terminalProcess.js#L18) (link is to the current `master` branch of VS Code as of the time when I'm writing this issue report), VS Code looks for the file `/usr/share/terminfo/x/xterm-256color`. If it is present, then the `name` variable is set to `xterm-256color`, otherwise it is set to `xterm`. It seems that `name` eventually ends up as the value of the `$TERM` environment variable.

And here is where there's a bit of a problem. On Debian, and distributions like Ubuntu and Mint that derive from Debian, there is no `/usr/share/terminfo/x/xterm-256color` file. The Debian package that contains `xterm-256color` puts it in `/lib/terminfo/x/xterm-256color` instead. (There is also a `/usr/share/terminfo/x/xterm+256color` file, but that one is not guaranteed to be installed on all systems since it is not marked as "essential". But the Debian `ncurses-base` package, which is marked as "essential" and therefore guaranteed to be installed on all systems, installs the `/lib/terminfo/x/xterm-256color` file).

Therefore, since `/usr/share/terminfo/x/xterm-256color` doesn't exist, when VS Code opens an integrated terminal it sets `$TERM` to be `xterm` instead of `xterm-256color`. And then Debian's `.bashrc` says "Oh, this isn't a color-capable terminal, better not use the color prompt".

**Note about Linux Mint:** In Linux Mint 18, the `.bashrc` has been modified to also use a color prompt if `$TERM` is set to `xterm`. Therefore, also Visual Studio Code still sets `$TERM` to `xterm`, incorrectly stating that it is not color-capable, Mint 18 actually *ignores* this and uses a color prompt *anyway*. However, other programs run from inside the integrated terminal may see the incorrect value of `$TERM` and choose not to display color when they could have, so it's still worth fixing this for Mint as well. Thankfully, Mint places its `xterm-256color` file in the same place as Debian and Ubuntu, so the suggested fix is the same for all of these distributions.

## Suggested fix

The `fs.existsSync('/usr/share/terminfo/x/xterm-256color')` line in `src/vs/workbench/parts/terminal/electron-browser/terminalProcess.js` could instead be changed to check for two files instead of just one:

    name = (fs.existsSync('/usr/share/terminfo/x/xterm-256color') || fs.existsSync('/lib/terminfo/x/xterm-256color')) ? 'xterm-256color' : 'xterm';

I have tested this change on Ubuntu 14.04 (Trusty) and it has the desired effect: inside the integrated terminal, `$TERM` is now set to `xterm-256color` and the Bash prompt is now colored instead of black&white. I have not tested this on the other affected distributions, but there is no reason why it should not work there as well.

**Comment by rmunn** [comment]
**Update**: On further testing, I learned that Debian 8 (Jessie) has a `.bashrc` that **only** sets up a color prompt if the value of `$TERM` is precisely `xterm-color`. Debian unstable (Sid) also has the same `.bashrc` behavior: it wants a value of `$TERM` that is specifically `xterm-color` and nothing else; a value of `xterm-256color` will not produce a color prompt. Both Ubuntu and Mint expand `.bashrc` to accept `xterm-256color` as producing a color prompt, but Debian (as of this writing) does not.

Therefore, if a color prompt is desired on Debian systems, it might be necessary to use the `lsb_release` command, with either the `-cs` option (producing "jessie" or "sid" as output) or the `-rs` option (producing "8.4" or "unstable" as output, though the "8.4" number will, of course, increase as new point releases of Debian Jessie are released). Then, if Debian is detected, the `name` variable (which becomes the value of `$TERM`) could be set to `xterm-color` instead of `xterm-256color`. This would come at the cost of running an extra process before being able to show the terminal, but it's the only way I've found that is guaranteed to work -- relying on the `/etc/debian_version` file will, unfortunately, produce false positives on Ubuntu (the Ubuntu Trusty system I'm typing this on contains "`jessie/sid`" in its `/etc/debian_version` file, which is a lie: this is not, in fact, Jessie, and `lsb_release -cs` correctly reports `trusty`).

**Comment by Tyriar** [comment]
Interesting, I always had issues with the color prompt initialization on Ubuntu so I removed it in my `.bashrc`. I'm quite surprised that this hasn't been reported before though if it does indeed disable color prompts by default.

Do you think just changing it to `xterm-256color`, without doing the file check would be sufficient? I'm not entirely sure if it's useful at this point; surely a machine that runs VSCode would be capable of displaying a color prompt?

**Comment by rmunn** [comment]
A bit of research suggests that just about *everything* supports 256 colors, so defaulting to `xterm-256color` should work just about everywhere. Except, surprisingly enough, the Ubuntu-on-Windows terminal, which only supports 16 colors -- at least according to [this bug report](https://github.com/fish-shell/fish-shell/issues/2910) from April 2016. But anyone running VS Code on Windows is probably not going to run it from the Ubuntu-on-WIndows terminal, so that shouldn't be an issue.

Also, setting `$TERM` to `xterm-256color` will fail on Debian Jessie and Sid as I mentioned in my earlier comment, but that's not really something that VS Code can do anything about. If the integrated terminal supports 256 colors, it should *report* that it supports 256 colors, and `xterm-256color` is the right way to report that. The fact that the Debian color prompt doesn't check for the name `xterm-256color` is a bug in Debian, not in VS Code.

The other place where setting `$TERM` to `xterm-256color` might cause problems is if people are SSHing into *really old* Unix systems, whose terminfo files have definitions for `xterm` but not for `xterm-256color`. But I don't think that's going to be a common usage scenario for VS Code. Most people starting an SSH session from the VS Code terminal are probably going to be doing so in order to copy files to/from a Web server, either to deploy their just-edited files or to fetch the current set of files in order to edit them. And that's going to be SCP or SFTP, where the value of `$TERM` won't matter. And even if someone *was* doing an SSH session into a remote machine with no support for `xterm-256color`, there would be no way for VS Code to find out about that. The file system check can only tell if the user's local machine knows about `xterm-256color`; there's no way to check that every single possible machine he might SSH into will know about it.

So on balance, I agree that it should be safe to just use `xterm-256color` without checking the filesystem.

**Comment by rmunn** [comment]
Speaking of `.bashrc`, here are the changes I made to my `.bashrc` on Ubuntu 14.04 to make the color terminal work:

```diff
diff -u /etc/skel/.bashrc ~/.bashrc
--- /etc/skel/.bashrc	2014-04-09 08:03:15.000000000 +0700
+++ /home/rmunn/.bashrc	2016-01-21 10:45:09.647534332 +0700
@@ -35,9 +35,15 @@
     debian_chroot=$(cat /etc/debian_chroot)
 fi
 
+# GNOME Terminal is broken re: $TERM settings
+if [ "$COLORTERM" = "gnome-terminal" -a "$TERM" = "xterm" ]; then
+    TERM="xterm-256color"
+fi
+
 # set a fancy prompt (non-color, unless we know we "want" color)
 case "$TERM" in
-    xterm-color) color_prompt=yes;;
+    xterm-color|xterm-256color) color_prompt=yes;;
+    *-256color) color_prompt=yes;;
 esac
 
 # uncomment for a colored prompt, if the terminal has the capability; turned
```

So the `xterm-256color` value still might not work on Ubuntu 14.04 (Trusty), because the *default* `.bashrc` on Ubuntu Trusty only checks for `xterm-color`. But that's a bug in Ubuntu.

However, I've now identified *two* systems that may still be in common use -- Ubuntu Trusty, which some people may still be using. (For example, I'm using it as a development machine right now, because I'm working on a Web app that's deployed to a server that runs Trusty, and the people managing that server have no intention of upgrading to Xenial before next year at the earliest). So I see two options:

1) Set the `$TERM` to `xterm-color` instead of `xterm-256color`. This is, IMHO, the wrong approach, because it lies about the integrated terminal's capabilities -- and the bug in Debian and Ubuntu will eventually be fixed. (It was indeed fixed in Ubuntu 16.04 (Xenial), where the `.bashrc` checks for either `xterm-color` or `*-256color` to set the color prompt). Or:

2) Set the `$TERM` to `xterm-256color` and add a note in a FAQ or in the integrated terminal's instructions, saying "If you were expecting your prompt to be in color and it isn't, here's how to fix it: look in your `.bashrc` file for the `case "$TERM" in` line, and change the `xterm-color` case to `xterm-color|*-256color`." I'm going to submit a bug report to Debian if there isn't one already, so hopefully Debian's `.bashrc` will also be fixed. This, IMHO, is the right approach, because it correctly reports VS Code's terminal's capabilities, and treats "My distro doesn't understand `xterm-256color`" as a bug in the distro that will eventually be fixed.

**Comment by rmunn** [comment]
**UPDATE:** Debian Sid has recently fixed the bug I was about to report where `xterm-256color` wasn't recognized. So setting the value of `$TERM` to `xterm-256color` is the right thing to do everywhere: if someone is running an older distro that doesn't set a color prompt when it sees that value, that bug will eventually be fixed when they upgrade their distro.

TL;DR: **Just set `xterm-256color`.** It's supported everywhere that VS Code can run. There will still be a few distros that won't set a color prompt when they see that `$TERM` value, but those distros do support the value, they just don't realize that it means a color prompt is possible. So add a FAQ to tell people what to set in their `.bashrc` (add the `*-256color` case) if they're running an older distro.

**Comment by Tyriar** [comment]
👍 feel free to put out a pull request if you can, otherwise I'll probably get to this tomorrow and follow up with an FAQ entry. Thanks a lot for the thorough research 😃

**Comment by Tyriar** [comment]
I pushed a fix for this as tomorrow will only accept critical issues and this is relatively minor. Created https://github.com/Microsoft/vscode-docs/issues/482 to follow up FAQ entry.
